### ISSUE TITLE
Write `CONTRIBUTING.md`

### ISSUE THREAD

**Issue by jswrenn** [rust-itertools/itertools#756]
Itertools needs a `CONTRIBUTING.md`. Key highlights should include:
- Picking the right receiver. (see #710)
- Testing (especially property-based testing).
- How to implement and robustly test `size_hint`.

A complete guide should also address:
- When to implement `DoubleEndedIterator`
- When to implement `ExactSizeIterator`
- When to implement `FusedIterator`.

(Anything else?)

Happy to provide mentorship on this.

**Comment by Philippe-Cholet** [comment]
I recently contributed about each of those points (except `DoubleEndedIterator`) and there are key points I would have liked to know early on so I think I can help here. _I'm starting a draft:_ https://github.com/Philippe-Cholet/itertools/compare/master...contributing
Mentorship welcomed.

**Comment by Philippe-Cholet** [comment]
- I would like some feedback on my [draft](https://github.com/Philippe-Cholet/itertools/blob/contributing/CONTRIBUTING.md) of "CONTRIBUTING.md".
- However, even if I am familiar with all kind of tests we do here, I did not write anything about testing yet as it's still unclear to me what should be suggested exactly. I think some clarification would be useful.
- After seeing "specialization tests" a few days ago, I had ideas to extend this further. I'd like to know if such thing would be of interest (we would discuss this further in another issue):

`tests/specializations.rs` definitely seems promising (I really like the `Unspecialized` struct) but all apply some "closure" **only on a fresh iterator** and _not on an iterator that has yielded elements_. A buggy iterator could behave differently if those were applied later on.
I've experimented quite a bit (and eventually ditched macros, for sanity) to check `size_hint/count/last/nth/fold/collect` (as most of our specialized methods are those) at every step of iteration (at the start and after every `next`). And it works fine but is slower as you might expect, so I added the possibility to limit this to a few steps.
Usage ([example](https://github.com/rust-itertools/itertools/commit/8c74f3331442b89c47185110e06cc5b0d9b12171)) would remain nice but my implementation is definitely way more verbose (working branch: https://github.com/rust-itertools/itertools/compare/master...Philippe-Cholet:itertools:test-specializations).
